lib/ecto/association.ex

```diff
@@ -587,16 +587,15 @@ defmodule Ecto.Association.Has do
     cardinality = Keyword.fetch!(opts, :cardinality)
     related = Ecto.Association.related_from_query(queryable, name)
 
-    ref =
-      module
-      |> Module.get_attribute(:primary_key)
-      |> get_ref(opts[:references], name)
+    ref_option = opts[:references]
 
-    unless Module.get_attribute(module, :ecto_fields)[ref] do
-      raise ArgumentError, "schema does not have the field #{inspect ref} used by " <>
+    if ref_option && Module.get_attribute(module, :ecto_fields)[ref_option] == nil do
```

Comment:
Moved this check up before the call to get the module's primary_key since what is returned by `Module.get_attribute(module, :ecto_primary_keys)` will for sure be in `:ecto_fields`